Mapping Issues Export

DerekBev opened the pull request:
@zlavergne 
This includes an npm-shrinkwrap.json file in the client directory. Since it wasn't there before, I don't think it's actually necessary, but I had to add it to get the project to build on my computer.

I've re-cloned the repo with all my changes and built again from the bottom up to make sure it all works, and it does.

DerekBev committed:
Finally working

Needs node 11.15.0 (use nvm) and gulp 3.9.1 to build

DerekBev committed:
Added mapping issue csv button functionality

DerekBev committed:
Added button for retrieving mapping issues as csv

DerekBev committed:
Added method for mapping issues csv functionality

DerekBev committed:
Added api for retrieving mapping issues as csv

DerekBev committed:
Added get_all_issues() and get_issue_row_count() to TaskMappingIssue class

DerekBev committed:
Added service for retrieving all mapping issues as csv

DerekBev committed:
Added endpoint for mapping issues csv api

DerekBev committed:
Improved whitespace in MappingIssuesApi

DerekBev committed:
Added logic to return mapping issues as .csv

DerekBev committed:
Changed method name from get_all_mapping_isues to get_mapping_issues

DerekBev committed:
Added project_id to MappingIssuesAPI

DerekBev committed:
Added parameter 'detailed_view' to MappingIssuesAPI

DerekBev committed:
Refactored, commented, and fixed a bug

DerekBev committed:
Added argument <detailed_view> to MappingIssuesAPI url

DerekBev committed:
Added styling for the detailed view checkbox

DerekBev committed:
Added detailedIssues functionality

DerekBev committed:
Added mapping issues button and checkbox

DerekBev committed:
Added mapping issue button functionality

DerekBev committed:
Removed mapping issues button from mapping issues page

DerekBev committed:
Added numpy

DerekBev committed:
Replaced numpy version with a version that works

DerekBev committed:
Matched code style

DerekBev committed:
Added hiding zero-only rows functionality

tsmock commented on client/package.json:
```diff
@@ -62,9 +64,9 @@
   },
-  "dependencies": {}
+  "dependencies": {
+    "node": "^11.15.0"
```
This is (pretty much) a deal breaker right now.

The version of node on the server is v8. While this PR _might_ work on the server, it would be appreciated if you could test with the server version (the way node was installed on the server makes it a PITA to update).

tsmock commented on server/models/postgis/task.py:
```diff
@@ -134,8 +134,13 @@ def as_dto(self):
         issue_dto.count = self.count
         return issue_dto
 
-    def __repr__(self):
```
Why did you remove this `__repr__` method?

DerekBev replied:
With an approach I was trying at first, I needed mapping issues to be returned from the database as objects so I could access their members, including mapping_issue_category_id. (I was using the two methods I wrote here, so issues were only returned as the string "issue: count".) There were no calls to any methods that used this __repr__ method, so I removed it.

Now I've implemented the functionality a different way, so it makes no difference for the issues export functionality. I've realized all of the methods here (__repr__, get_all_issues(), get_issue_row_count()) are now unused, so I plan to remove the two I wrote. I can add the __repr__ method back in if you'd like, but it's still unused.

tsmock replied:
This `__repr__` method is in v3.4.9 of the TM. I would leave it in.

tsmock commented on server/services/mapping_issues_service.py:
```diff
@@ -45,3 +53,219 @@ def delete_mapping_issue_category(category_id: int):
+        if (zerosRows == "true"):
+            zeros = True
+        
+        #user_dict {str username : {int task (taskId) : {str issue : str issue_count}}}
```
Why did you leave this in the code? You probably should have either:
a) Explained why the code was commented out or
b) Removed the commented out code (preferred)

DerekBev replied:
This isn't actual code. It's comments describing what the return data types are for the following functions. I wrote these and left them because, if I were to modify anything in the get_mapping_issues method in the future, I would have to dig through what I wrote, figure out what everything is, and write them out again so I would know what I was working with. I can change them to multi-line string comments so it's clear it's not code.

tsmock replied:
These would have been better documented in `MappingIssueExporter.compile_validated_tasks_by_user`.

For example:
```diff
diff --git a/server/services/mapping_issues_service.py b/server/services/mapping_issues_service.py
index b458b550..385847a4 100644
--- a/server/services/mapping_issues_service.py
+++ b/server/services/mapping_issues_service.py
@@ -8,6 +8,7 @@ from server.models.dtos.stats_dto import ProjectContributionsDTO
 from server.services.stats_service import StatsService
 from copy import deepcopy
 import numpy as np
+import typing
 
 class MappingIssueCategoryService:
 
@@ -182,7 +183,13 @@ class MappingIssueExporter:
 
 
     @staticmethod
-    def compile_validated_tasks_by_user(project_id: int, zeros: bool):
+    def compile_validated_tasks_by_user(project_id: int, zeros: bool) -> typing.Tuple[typing.Dict[str, typing.Dict[int, typing.Dict[str, str]]], typing.Set[any], int, typing.Dict[int, any]]:
+        """
+        :returns: user_dict: {str username: {int task (taskId) : {str issue : str issue_count}}},
+            project_users: Set[str: users on project]
+            num_validated_tasks: int
+            tasks_as_tasks_dict: {int taskId: task }
+        """
         all_project_tasks = Task.get_all_tasks(project_id)
         validated_tasks = []
         project_users = []
```

You'll want to double-check that, but it should be good for python 3.6. And you will want to replace the `any` with what it is supposed to be (`UserDTO` or `str`, or whatever).

tsmock replied:
Note: As a bonus, if you use add the typing, most python IDEs will understand it, and give usable method possibilities.

tsmock reviewed (changes requested):
I haven't done a thorough review (I just skimmed looking for glaring issues).

Notes:
* Please keep in mind that we have been meaning to migrate to TM4. My general recommendation is to get code upstreamed prior to making a PR with the code here (this way, we (a) don't have to maintain the code as much going forward and (b) we don't have to do as much work rebasing our branch in the future).
* TM3 is a decently old codebase (like I said elsewhere, the server is running node v8, which we should probably look at updating soonish, but will probably be done in step with the update to TM4)
* Please run `black`/`flake8` and fix the issues that pop up (use the versions specified in `requirements.txt` -- this will help us have a smaller diff w.r.t. v3.4.x to look at when moving to TM4). I haven't checked to see if you have run `black` or `flake8`, so this is just a general recommendation.

tsmock commented:
I've made (most) of my recommended changes in https://github.com/KaartGroup/tasking-manager/tree/mapping-issues-export .

Do you have a series of tests for the functionality?

tsmock commented on server/services/mapping_issues_service.py:
```diff
@@ -45,3 +53,219 @@ def delete_mapping_issue_category(category_id: int):
+    def compile_validated_tasks_by_user(project_id: int, zeros: bool):
+        all_project_tasks = Task.get_all_tasks(project_id)
+        validated_tasks = []
+        project_users = []
```
If this is supposed to be a set, you may want to use `set()`. This is unordered, so if it has to be a ordered, then ignore this.

See https://docs.python.org/3/library/stdtypes.html#set-types-set-frozenset for more information.

DerekBev committed:
Moved comments and made minor changes

DerekBev committed:
Reverted altered dependency versions for testing

DerekBev committed:
Removed npm-shrinkwrap.json as it's not needed

tsmock commented:
Ping?

DerekBev committed:
Merge pull request #2 from DerekBev/issues-export-test

Issues export test

DerekBev commented:
Soo... I'm planning to get it upstreamed first as you suggested. I hadn't realized what I was writing could be used with hotosm/tasking-manager (I thought Kaart's version was different somehow.), so I'm adding some functionality to make it usable in all conditions and more of a finished product. (Things I wanted to do anyways. I'm almost done with this part)

I'm also writing automated tests. I just had a list of things I was manually testing, but it really needs unit tests added to the suite, so I'm working on that. 

I'll get back to you when these things are done.

It works with Node 8, so I removed that dependency. It also works with all the versions originally listed in requirements.txt, so I've changed those back. Some things didn't work on my machine with those versions before, but now they do. I don't know what changed.

Black wants to change a lot. In a lot of files. It looks like it hasn't been used on our fork of the TM, yet, at least not on the files I checked. I've looked at the errors from flake8, but haven't changed much, yet. I'm planning to do these things last.

DerekBev committed:
Exclude archived issues

DerekBev committed:
Added formatting for commas in fields

DerekBev committed:
Renamed MappingIssueExporter to MappingIssueExportService

DerekBev committed:
Added unit tests for issues export

DerekBev committed:
Add tearDownClass() method

DerekBev committed:
Fix tearDownClass() method

DerekBev commented:
After taking a look at the frontend of the upstream repo, I've realized it has been completely redesigned and refactored and now uses react. Since any changes to this code beyond this point won't make our diffs any smaller when moving to TM4, I don't see much benefit in waiting any longer to merge the code here. I can still get the code upstreamed, but it will take more time. All the issues above have been resolved.

DerekBev commented:
Ping?

tsmock commented:
Would you mind running `python3 manage.py refresh_translatables`? Its causing a test to fail ( see https://gitlab.com/gokaart/tasking-manager/-/jobs/1057321923#L438 ).

DerekBev committed:
Refreshed translatables

DerekBev commented:
Done. Let me know if there's anything else.

tsmock commented:
Looks good to me. I've squished the commits and made a new PR over here: https://github.com/KaartGroup/tasking-manager/pull/61. Please check to make certain that everything works as _you_ expected it to.

DerekBev commented:
I just did a final check and everything works as expected.
